src/encode/pattern/mod.rs

```diff
@@ -536,6 +545,47 @@ impl<'a> From<Piece<'a>> for Chunk {
                         params: parameters,
                     }
                 }
+                #[cfg(feature = "log_kv")]
+                "K" | "key_value" => {
+                    if formatter.args.len() > 2 {
+                        return Chunk::Error("expected at most two arguments".to_owned());
+                    }
+
+                    let key = match formatter.args.first() {
+                        Some(arg) => {
+                            if let Some(arg) = arg.first() {
+                                match arg {
+                                    Piece::Text(key) => key.to_owned(),
+                                    Piece::Error(ref e) => return Chunk::Error(e.clone()),
+                                    _ => return Chunk::Error("invalid log::kv key".to_owned()),
+                                }
+                            } else {
+                                return Chunk::Error("invalid log::kv key".to_owned());
+                            }
+                        }
+                        None => return Chunk::Error("missing log::kv key".to_owned()),
+                    };
+
+                    let default = match formatter.args.get(1) {
+                        Some(arg) => {
+                            if let Some(arg) = arg.first() {
+                                match arg {
+                                    Piece::Text(value) => value.to_owned(),
+                                    Piece::Error(ref e) => return Chunk::Error(e.clone()),
+                                    _ => return Chunk::Error("invalid log::kv default".to_owned()),
+                                }
+                            } else {
+                                return Chunk::Error("invalid log::kv default".to_owned());
+                            }
+                        }
+                        None => "",
+                    };
```

Comment:
Id like to see this and MDC stuff get factored into a helper function rather than a copy-paste.